src/documents/posts/2015-01-16-commonjs-modules-make-brittle-singletons.md

```diff
@@ -0,0 +1,68 @@
+---
+title: CommonJS Modules Make Brittle Singletons
+author: Adam Hull
+layout: post
+disqus:
+  shortname: goodeggsbytes
+  url: "http://bites.goodeggs.com/post/commonjs-modules-make-brittle-singletons"
+style: |
+  .full-width img {
+    width: 100%;
+    padding: 0;
+    margin-bottom: 1em;
+  }
+---
+
+We ocassionally rely on node's [module caching](http://nodejs.org/api/modules.html#modules_caching) to share a single instance throughout a full-stack javascript project. This strategy breaks more than we'd like.
+
+If modules `butternut` and `delicata` both `require('squash')`, they'll usually get the same (think `===`) squash instance. But not always.
+
+Here are a couple times it hasn't worked out.
+
+<!-- more -->
+
+<div class="full-width">
+![Olive &amp; Sinclair Chocolate Co
+Bourbon Nib Brittle](/images/posts/commonjs-modules-make-brittle-singletons/brittle.jpg)
+</div>
+
+
+Let's say we're really into [node-fibers](https://github.com/laverdet/node-fibers) with its concise coroutines and error handling. We're starting a new module, tests first, so we `npm install mocha mocha-fibers` and write some failing tests.  Next we [`npm install fibrous`](https://github.com/goodeggs/fibrous) to help implement our module.  If we list installed fibers with `npm ls fibers` we get:
+
+```
+├─┬ fibrous@0.3.3
+│ └── fibers@1.0.4
+└─┬ mocha-fibers@1.1.0
+  └── fibers@1.0.4
+```
+
+Uh oh, we've got two! This can cause some pretty [wierd](https://github.com/laverdet/node-fibers/issues/102) [behavior](https://github.com/laverdet/node-fibers/issues/177).  Luckily, fibers was [patched](https://github.com/laverdet/node-fibers/commit/d9bc3a7b9d486d6f45170501de8626d52dfa5dfa) in 1.0.4 to mitigate  this particular problem using global variables.  If we care  which version of `fibers` our project uses, it's best to install it explicitly as a top level dependency, before installing `fibrous` or `node-fibers`.  For example, `npm install fibers fibrous mocha-fibers` yields:
+
+```
+├── fibers@1.0.5
+├── fibrous@0.3.3
+└── mocha-fibers@1.1.0
+```
+
+Only one `fibers`, much better.
+
+Recap: multiple modules with a shared dependency can get different instances of that dependency. Nothing super new there, there's even a [big caveat in the docs](http://nodejs.org/api/modules.html#modules_module_caching_caveats) about it.
```

Comment:
Though it seems like peer dependencies are on their way out:
https://github.com/npm/npm/issues/6565
On Jan 16, 2015 3:16 PM, "Alon Salant" <notifications@github.com> wrote:

> In
> src/documents/posts/2015-01-16-commonjs-modules-make-brittle-singletons.md
> <https://github.com/goodeggs/bites/pull/48#discussion-diff-23116583>:
>
> > +│ └── fibers@1.0.4
> > +└─┬ mocha-fibers@1.1.0
> > +  └── fibers@1.0.4
> > +```
> > +
> > +Uh oh, we've got two! This can cause some pretty [wierd](https://github.com/laverdet/node-fibers/issues/102) [behavior](https://github.com/laverdet/node-fibers/issues/177).  Luckily, fibers was [patched](https://github.com/laverdet/node-fibers/commit/d9bc3a7b9d486d6f45170501de8626d52dfa5dfa) in 1.0.4 to mitigate  this particular problem using global variables.  If we care  which version of `fibers` our project uses, it's best to install it explicitly as a top level dependency, before installing `fibrous` or `node-fibers`.  For example, `npm install fibers fibrous mocha-fibers` yields:
> > +
> > +```
> > +├── fibers@1.0.5
> > +├── fibrous@0.3.3
> > +└── mocha-fibers@1.1.0
> > +```
> > +
> > +Only one `fibers`, much better.
> > +
> > +Recap: multiple modules with a shared dependency can get different instances of that dependency. Nothing super new there, there's even a [big caveat in the docs](http://nodejs.org/api/modules.html#modules_module_caching_caveats) about it.
>
> There's another approach from npm - peer dependencies. The discussion of
> peer dependencies at http://blog.nodejs.org/2013/02/07/peer-dependencies/
> touches on the same problem as this blog post.
>
> Peer dependencies work great if the module author uses them - for example,
> mocha-fibers could require node-fibers as a peer dependency to avoid the
> issue you ran in to.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/goodeggs/bites/pull/48/files#r23116583>.
>